### PR TITLE
Kat Edges 

### PR DESCRIPTION
**Pull request by kat-perreira**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In the model for user, I wrote a method that returns an array of works voted on by itself
Describe how you approached testing that model method. What edge cases did you come up with? | I tested the functionality of it, I guess an edge case would be the validation test that ensures that a user must be logged in to vote but I didn't do any edge case tests for my business model methods, just checked that they worked.
What are `session` and `flash`? What is the difference between them? |  session: hash like object provided by rails to store stuff, in my program I used it for the logging in/out feature. flash: also a hash like object to store messages. Flash is temporary, only present for one request/response cycle, sessions is long term.
Describe a controller filter you wrote. |  I wrote one control filter, to find the logged in user. In app/controllers/application_controller.rb,  I wrote a private method that returns an instance variable to find a logged in user in order to DRY up some code
What was one thing that you gained more clarity on through this assignment? | Everything but If I have to pick just one i'd say debugging.
What is the Heroku URL of your deployed application? | https://media-ranker-kat-edges.herokuapp.com  UPDATE: Figured it out. I thought I could deploy and look at it empty before seeding the data, I was wrong. The link should work now.  

Do you have any recommendations on how we could improve this project for the next cohort? |  I think this is a really good project, I learned a lot and it showed me areas I need better understandings of things. One thing i've had to do twice now is deal with authentication token/turbo links errors, maybe a few words about that? I am not sure what it does exactly, just that it has to do with loading speed and taking out a line in one of the rails generated folders eliminates the authentication error Ive always gotten.

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Comprehension questions	| x
**General** |
Rails fundamentals (RESTful routing, use of named paths) | x
Views are well-organized (DRY, use of semantic HTML, use of partials) | x
Errors are reported to the user | x
Business logic lives in the models | x
Models are thoroughly tested, including relations, validations and any custom logic | good test coverage on validations and relations, was hoping for a little more coverage on custom logic
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x, deleting a work was broken for me
**Wave 2 - Users and Votes** |
Users can log in and log out | x
The ID of the current user is stored in the session | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count |
Splash page contains a media spotlight |
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | x
Individual user pages and the user list are present |
**Optional - Styling** |
Bootstrap is used appropriately | x
Look and feel is similar to the original | largely was met, details were missed (flash messages)
**Overall** |

Good work Kat-- overall, you have an interesting, multi-functional app that looks pretty good!

Overall, my comments on your code otherwise will have the following themes:
- Tiny places to clean up that are moreso suggestions than anything else, but will make your code better
- Some questions on some of the custom model methods and how they're implemented and tested

Also in general, you have a few pieces of functionality missing:
- User pages-- individual ones and listing all users
- Sorting works by vote, getting the top ten per category or getting the overall top one

I'm very happy you got logging in and logging out though! In general, I think you do a great job of getting the systems thinking and the overall user flow, and there are little bits of custom method implementation to review.

Good work overall!

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'homepages#index'
+
+  resources :votes, :works
```
Comment (tildeee) [inline config/routes.rb:6]: `resources` produces every RESTful route for this resource. Do we need every RESTful route for votes?

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  resources :works do
+    resources :votes
+  end
+
+  resources :users do
```
Comment (tildeee) [inline config/routes.rb:12]: do we need edit/delete routes on users?

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    resources :votes
+  end
+
+
+  resources :homepages, only: [:index]
+
+  resources :sessions, only: [:new, :create]
+  post '/sessions/logout', to: 'sessions#logout', as: 'logout'
```
Comment (tildeee) [inline config/routes.rb:20]: These custom routes look great!

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+class User < ApplicationRecord
+  has_many :votes
+  validates :username, presence: true, uniqueness: true
+
+
+  # create a vote object with current user's id and the work id passed in
+  def vote(work_id)
+    vote = Vote.where(user_id: self.id, work_id: work_id)
+    return vote
```
Comment (tildeee) [inline app/models/user.rb:9]: Your comments say that this "create"s a vote object, but it calls `Vote.where`... it doesn't create a Vote, it just finds any existing votes, and returns it as an array

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  # Assign the list of votes to the variable votes_array...helper method
+  def users_votes
+    votes_array = Vote.where(user_id: self.id)
+    return votes_array
+  end
```
Comment (tildeee) [inline app/models/user.rb:16]: Why not just use `.votes`?

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+  # collect the work ids for all the works that the user upvoted
+  def upvoted
+    votes = users_votes
+    upvoted_workids = []
+    votes.each do |vote|
+      upvoted_workids << vote
```
Comment (tildeee) [inline app/models/user.rb:23]: This loop looks through an array of votes and... for each vote... it shovels the vote into another array named `upvoted_workids`. It doesn't shovel in the vote's work id

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    return upvoted_workids
+  end
+
+  # return the works users voted on here, not in the view.
+
+  def user_voted_works
+    work_ids = upvoted
+    works = []
+    work_ids.each do |id|
+      work = Work.find_by(id: id)
```
Comment (tildeee) [inline app/models/user.rb:34]: This ends up working because Rails/Active Record ends up accepting an ActiveRecord object (a `Vote` object) and when you `find_by` something with `id` and pass in the `ActiveRecord` object, it'll go assume to search by the id. However, each element in the `work_ids` array is actually a `Vote` object.

File: app/models/work.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      user = User.find_by(id: num)
+      users << user
+    end
+
+    return users
+  end
+
+  def upvoted_works(user_id)
+    vote = Vote.where(user_id: user_id, work_id: self.id)
+    return vote
```
Comment (tildeee) [inline app/models/work.rb:33]: Should these be named like this? What does this method actually do? Is `upvoted_works` the best name for it? Does the local variable `vote` represent one vote? What does `Vote.where` return?

File: app/views/homepages/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,37 @@
+<article>
+  <h1> Media Spotlight</h1>
+    <%= link_to @media_spotlight.title, work_path(@media_spotlight.id) %>
+
+<div class="media_body">
+  <article class = "media_column">
```
Comment (tildeee) [inline app/views/homepages/index.html.erb:6]: Very small note: HTML convention is to not space that out so just `class="media_column"`

File: app/views/users/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+<%= render partial: "form" %>
+
+
+<h3>A Note About Logging In</h3>
+<p>
+  There is no password field. In fact, there is no indication whatsoever that you are who you say you are. There's nothing special about users - username is just another piece of data that the user entered and we have to keep track of.
+
+We'll learn more about security and authentication in the next couple weeks. For now, don't worry about it beyond what you can see here.
+</p>
```
Comment (tildeee) [inline app/views/users/new.html.erb:11]: Does this view get used?

File: app/views/works/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,45 @@
+
+
+  <strong>
+
+    <h2> Title: <%= @work.title %></h2>
+    <p> Created by: <%= @work.creator %></p>
+    <p> Published: <%= @work.publication_year %></p>
+    <p> Description: <%= @work.description %></p>
```
Comment (tildeee) [inline app/views/works/show.html.erb:8]: Why are these nested in a strong tag? If you wanted them to be bold, you shouldn't use HTML for that, you should use CSS!

File: test/models/user_test.rb [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+  # relation tests: a user has many votes
+  describe 'has relations' do
+    it 'a user can have votes' do
+      user = users.first
+      work_first = works.first
+      work_second = works.second
+
+      vote = Vote.create!(user_id: user.id, work_id: work_first.id)
+      vote = Vote.create!(user_id: user.id, work_id: work_second.id)
```
Comment (tildeee) [inline test/models/user_test.rb:13]: instead of reassigning the same variable `vote`, when all you really care about is making the Votes and not having a reference to them by some variable, you should feel free to write this instead:
```ruby
Vote.create!(user_id: user.id, work_id: work_first.id)
Vote.create!(user_id: user.id, work_id: work_second.id)
```

File: test/models/user_test.rb [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+      @test_work_1 = works(:Haunting)
+      @test_work_2 = works(:Shrek)
+      @test_vote_1 = Vote.create!(user_id: @test_user.id, work_id: @test_work_1.id)
+      @test_vote_2 = Vote.create!(user_id: @test_user.id, work_id: @test_work_2.id)
+    end
+
+    it '#vote(work_id) returns a vote object with a specific work id' do
+      vote = @test_user.vote(@test_work_1.id)
+      id = vote.first.id
+      vote_id = @test_vote_1.id
```
Comment (tildeee) [inline test/models/user_test.rb:67]: The line above proves that the method `vote(work_id)` doesn't return a vote object... it returns an array of votes. That's why you had to call `vote.first`

File: test/models/user_test.rb [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    it '#users_votes returns an array of votes' do
+      ids = @test_user.users_votes
+
+      expect (ids.count).must_equal 2
+
+    end
+
+    it '#upvoted returns an array of work ids' do
+      work_ids = @test_user.upvoted
+      expect(work_ids.length).must_equal 2
```
Comment (tildeee) [inline test/models/user_test.rb:81]: You didn't test that this array contained work ids, when in reality it returns `Vote` objects!

File: test/models/user_test.rb [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    end
+
+  end
```
Comment (tildeee) [inline test/models/user_test.rb:84]: Tests for `user_voted_works`?
